crates/functional-yew/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+        Box::new(|a: &T1, _: &(), _: &(), _: &(), _: &()| callback(a)),
+        o1,
+        (),
+        (),
+        (),
+        (),
+    );
+}
+
+pub fn use_effect2<F, Destructor, T1, T2>(callback: F, o1: T1, o2: T2)
+where
+    F: FnOnce(&T1, &T2) -> Destructor,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+{
+    let callback = Box::new(callback);
+    use_effect5(
+        Box::new(|a: &T1, b: &T2, _: &(), _: &(), _: &()| callback(a, b)),
+        o1,
+        o2,
+        (),
+        (),
+        (),
+    );
+}
+
+pub fn use_effect3<F, Destructor, T1, T2, T3>(callback: F, o1: T1, o2: T2, o3: T3)
+where
+    F: FnOnce(&T1, &T2, &T3) -> Destructor,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+    T3: PartialEq + 'static,
+{
+    let callback = Box::new(callback);
+    use_effect5(
+        Box::new(|a: &T1, b: &T2, c: &T3, _: &(), _: &()| callback(a, b, c)),
+        o1,
+        o2,
+        o3,
+        (),
+        (),
+    );
+}
+
+pub fn use_effect4<F, Destructor, T1, T2, T3, T4>(callback: F, o1: T1, o2: T2, o3: T3, o4: T4)
+where
+    F: FnOnce(&T1, &T2, &T3, &T4) -> Destructor,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+    T3: PartialEq + 'static,
+    T4: PartialEq + 'static,
+{
+    let callback = Box::new(callback);
+    use_effect5(
+        Box::new(|a: &T1, b: &T2, c: &T3, d: &T4, _: &()| callback(a, b, c, d)),
+        o1,
+        o2,
+        o3,
+        o4,
+        (),
+    );
+}
+
+pub fn use_effect5<F, Destructor, T1, T2, T3, T4, T5>(
+    callback: Box<F>,
+    o1: T1,
+    o2: T2,
+    o3: T3,
+    o4: T4,
+    o5: T5,
+) where
+    F: FnOnce(&T1, &T2, &T3, &T4, &T5) -> Destructor,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+    T3: PartialEq + 'static,
+    T4: PartialEq + 'static,
+    T5: PartialEq + 'static,
+{
+    struct UseEffectState<T1, T2, T3, T4, T5, Destructor> {
+        o1: Rc<T1>,
+        o2: Rc<T2>,
+        o3: Rc<T3>,
+        o4: Rc<T4>,
+        o5: Rc<T5>,
+        destructor: Option<Box<Destructor>>,
+    }
+    let o1 = Rc::new(o1);
+    let o2 = Rc::new(o2);
+    let o3 = Rc::new(o3);
+    let o4 = Rc::new(o4);
+    let o5 = Rc::new(o5);
+    let o1_c = o1.clone();
+    let o2_c = o2.clone();
+    let o3_c = o3.clone();
+    let o4_c = o4.clone();
+    let o5_c = o5.clone();
+    use_hook(
+        move |state: &mut UseEffectState<T1, T2, T3, T4, T5, Destructor>, hook_update| {
+            let mut should_update = !(*state.o1 == *o1
+                && *state.o2 == *o2
+                && *state.o3 == *o3
+                && *state.o4 == *o4
+                && *state.o5 == *o5);
+
+            if should_update {
+                if let Some(de) = state.destructor.take() {
+                    de();
+                }
+                let new_destructor = callback(
```

Comment:
If we are mirroring React, this callback should run on `mounted()`. I think `use_effect` in general is more complex than the other hooks, how about we split this out into a separate PR?

Reply:
In react the hook corresponds to componentDidMount, componentDidUpdate, and componentWillUnmount. DO you think it is currently run in the wrong spot / should it run after the render?

Reply:
If you wanted to use a ref to an element inside an effect, we should make sure the effect runs inside [`mounted()`](https://github.com/yewstack/yew/blob/master/src/html/mod.rs#L45) so that the ref is hydrated

Reply:
Oh right ofc. Yes that needs special handling... I think I know how though. We need a next_tick array, and if it's not empty, we send a message to ourselves right away.

Reply:
Cool, you will probably want to check out the [`Scheduler`](https://github.com/yewstack/yew/blob/master/src/scheduler.rs) as well

Reply:
Already did on my endavour to find a bug that didn't exist :D
It's crate-pub though so hooks need their own :)

Reply:
Done